### PR TITLE
Patches Master Build & Other Additions

### PR DESCRIPTION
**Pull request by GlitchOo**
**Pull Request Description**

-Fixes current master build with successful pnpm install && pnpm build
-Fixes a Native Execution error with the phone model and textures (kvpservice)
-Includes PR https://github.com/project-error/npwd/pull/997
-Adjusts the the position of the popper modal for images on the twitter app (centers on screen)
-Added image upload support for "upload.io"

**Pull Request Checklist**:
* [✓] Have you followed the guidelines in our Contributing document and Code of Conduct?
* [✓] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/project-error/npwd/pulls) for the same update/change?
* [ ✓] Have you built and tested NPWD in-game after the relevant change?

I've been testing it for over 48 hours now and everything is working perfectly fine

### CODE REVIEW

**Comment by AvarianKnight** [comment]
Going to wait to review this until after Mycrofts pr gets reviewed

**Comment by GlitchOo** [comment]
Added the ability to post twitter posts to discord via webhook with the hook set in a convar
Ex:
set TWITTER_WEBHOOK "URL"

Also added bool setting to twitter config that allows just an image to be posted to twitter without typing a message. (Disabled by default)

**Comment by Opy12345** [comment]
how can i download the phone?

**Comment by GlitchOo** [comment]
> how can i download the phone?

No official release of these changes yet as it's still pending a merge into the master branch. However I have made the last build of my changes available here: https://github.com/GlitchOo/npwd/releases/tag/1.7.2-G if you want them :)

**Comment by Opy12345** [comment]
> > how can i download the phone?
> 
> No official release of these changes yet as it's still pending a merge into the master branch. However I have made the last build of my changes available here: https://github.com/GlitchOo/npwd/releases/tag/1.7.2-G if you want them :)

ok so just install the zip file there?

**Comment by GlitchOo** [comment]
This latest push of changes does the following. All have been tested and are working.
- Enabled Dark chat (Fixed a crashing issue caused by missing svg icon and 
       patched failed to fetch members in group chat when offline
- Cleaned up contact lists (removed any code that wasn't needed or being used
- Made the call and text buttons functional on the contact view page
      https://gyazo.com/c1afa1c26ea17c5f1e267dc8688e1244
- Patched routing on the dialer app making contacts non responsive
      https://gyazo.com/c34a6ce352a7872144642de88a21a52e
- Fixed the position where the phone spawns in the players hands causing it to clip in hand
      https://i.gyazo.com/e19e4ad1b41b54b67f9e83f6537adfdc.png
- That's all I can think of that I've changed in the last commits since the last successful workflow tests

This should bring everything up to a fully functional phone and build again

**Comment by GlitchOo** [comment]
No idea when this will be approved and merged for the meantime you can get the unofficial build release here:
https://github.com/GlitchOo/npwd/releases/tag/1.7.3-G

**Comment by Opy12345** [comment]
> No idea when this will be approved and merged for the meantime you can get the unofficial build release here: https://github.com/GlitchOo/npwd/releases/tag/1.7.3-G

ok thank you.

**Comment by GlitchOo** [comment]
Okay this should be the last of the phones current functions now working.
https://gyazo.com/d946b201f8231e830688858a5088a769

**Comment by GlitchOo** [comment]
Added infinite scrolling to the twitter app
https://gyazo.com/63c2b2540fdd9729df14c92cf9bcd869
I guess all that's left is a merge into the master 👀👀👀

**Comment by ArielZ123** [comment]
> Added infinite scrolling to the twitter app https://gyazo.com/63c2b2540fdd9729df14c92cf9bcd869 I guess all that's left is a merge into the master 👀👀👀

Amazing let us know here when we can download it.

File: apps/game/server/twitter/twitter.service.ts [hunk 1]
```diff
@@ -6,12 +6,15 @@ import { reportTweetToDiscord } from '../misc/discord';
 import { PromiseEventResp, PromiseRequest } from '../lib/PromiseNetEvents/promise.types';
 import { getDefaultProfileNames } from '../players/player.utils';
 import { checkAndFilterImage } from './../utils/imageFiltering';
+import { tweetDiscordPost } from '../lib/http-service';
 
 class _TwitterService {
   private readonly twitterDB: _TwitterDB;
+  private readonly DISCORDWH: string;
```
Comment (AvarianKnight) [inline apps/game/server/twitter/twitter.service.ts:13]: Can you change this to `DISCORD_WEBHOOK` the abbreviation might confuse some

File: apps/phone/src/apps/contacts/components/views/ContactInfo.tsx [hunk 2]
```diff
@@ -81,10 +159,8 @@ const ContactsInfoPage: React.FC = () => {
 
   return (
     <div className="mx-auto h-full w-full">
-      <button
-        onClick={() => history.goBack()}
-        className="mt-4 ml-4 rounded-md px-3 py-1 hover:bg-gray-200 hover:dark:bg-neutral-800"
-      >
+      <SendMoneyModal open={contactPayModal} closeModal={() => setcontactPayModal(false)} openContact={number} />
```
Comment (AvarianKnight) [inline apps/phone/src/apps/contacts/components/views/ContactInfo.tsx:162]: This should probably be added as a config for the people who enable this integration

File: apps/phone/src/apps/contacts/hooks/useModal.tsx [hunk 3]
```diff
@@ -0,0 +1,8 @@
+import { useRecoilState } from 'recoil';
+import { showMoneyModal } from './state';
+
+export const useModal = () => {
+  const [contactPayModal, setcontactPayModal] = useRecoilState<boolean>(showMoneyModal);
```
Comment (AvarianKnight) [inline apps/phone/src/apps/contacts/hooks/useModal.tsx:5]: nit: this should be setContactPayModal to keep consistent with the rest of the project

File: apps/phone/src/apps/contacts/hooks/state.ts [hunk 4]
```diff
@@ -55,4 +61,4 @@ export const useContactsValue = () => useRecoilValue(contacts);
 export const useFilteredContacts = () => useRecoilValue(filteredContacts);
 
 export const useContactFilterInput = () => useRecoilState(filterInput);
-export const useSetContactFilterInput = () => useSetRecoilState(filterInput);
+export const useSetContactFilterInput = () => useSetRecoilState(filterInput)
```
Comment (AvarianKnight) [inline apps/phone/src/apps/contacts/hooks/state.ts:64]: Why remove the semicolon ☹️

File: apps/phone/src/apps/twitter/components/AddTweetModal.tsx [hunk 5]
```diff
@@ -80,8 +80,15 @@ const AddTweetModal = () => {
...
-    if (cleanedMessage.length === 0) return;
-    if (!isValidMessage(cleanedMessage)) return;
+    if (cleanedMessage.length === 0) {
+      if (!allowNoMessage || !images || images.length === 0)
+      {
+        return
+      }
+    }
+
+    if (cleanedMessage.length > 0)
```
Comment (AvarianKnight) [inline apps/phone/src/apps/twitter/components/AddTweetModal.tsx:90]: Why did you remove the return here?

File: apps/phone/src/apps/twitter/components/tweet/TweetList.tsx [hunk 6]
```diff
@@ -2,20 +2,47 @@ import { memo, useState } from 'react';
...
+import { processTweet } from '@apps/twitter/utils/tweets';
 
 export function TweetList({ tweets }: { tweets: FormattedTweet[] }) {
   const [imageOpen, setImageOpen] = useState<string | null>(null);
+  
+  const [tweetsData, setTweetsData] = useState(tweets);
+  const [page, setPage] = useState(0)
+  const [hasMore, setHasMore] = useState((tweetsData.length >= 20 ? true : false))
+
+  const loadmore = () => {
```
Comment (AvarianKnight) [inline apps/phone/src/apps/twitter/components/tweet/TweetList.tsx:19]: This doesn't solve the original issue the phone had with InfiniteScroll before, reference: https://github.com/project-error/npwd/issues/639#issuecomment-1050162876

Any infinite scrolling that would be added would need to be able to cull elements which no major library currently offers.

File: apps/game/server/contacts/contacts.controller.ts [hunk 7]
```diff
@@ -1,7 +1,29 @@
...
+import { config } from '../config';
+import { Contact, ContactDeleteDTO, ContactEvents, PreDBContact, ContactPay } from '@typings/contact';
+import PlayerService from '../players/player.service';
 import ContactService from './contacts.service';
 import { contactsLogger } from './contacts.utils';
 import { onNetPromise } from '../lib/PromiseNetEvents/onNetPromise';
+const exps = global.exports;
+
+onNetPromise<ContactPay, Contact>(ContactEvents.PAY_CONTACT, (reqObj, resp) => {
+  PlayerService.getIdentifierByPhoneNumber(reqObj.data.number, true).then((response) => {
```
Comment (AvarianKnight) [inline apps/game/server/contacts/contacts.controller.ts:10]: This should be behind an optional config

**Review by AvarianKnight — CHANGES_REQUESTED** [review]

File: apps/phone/src/apps/contacts/components/views/ContactInfo.tsx (hunk 2 above)
Comment (GlitchOo) [reply apps/phone/src/apps/contacts/components/views/ContactInfo.tsx:162]: it already is

File: apps/game/server/contacts/contacts.controller.ts (hunk 7 above)
Comment (GlitchOo) [reply apps/game/server/contacts/contacts.controller.ts:10]: again this already is from the client side

File: apps/phone/src/apps/contacts/components/views/ContactInfo.tsx [hunk 8]
```diff
@@ -73,6 +141,16 @@ const ContactsInfoPage: React.FC = () => {
     updateContact({ id: contact.id, number, avatar, display: name });
   };
 
+  const openpayModal = () => {
+    if (ResourceConfig.general && ResourceConfig.general.useResourceIntegration && 
+        ResourceConfig.contacts && ResourceConfig.contacts.frameworkPay) {
```
Comment (AvarianKnight) [inline apps/phone/src/apps/contacts/components/views/ContactInfo.tsx:146]: Couldn't you use option chaining here? Feel free to correct me if I'm wrong
`ResourceConfig?.general?.useResourceIntegration`

File: apps/phone/src/apps/twitter/components/AddTweetModal.tsx (hunk 5 above)
Comment (GlitchOo) [reply apps/phone/src/apps/twitter/components/AddTweetModal.tsx:90]: I didn't, I modified it to to check to see if they have enabled the new setting in the twitter section of the config (disabled by default) to check if they wish to let users post only an image with no message. Otherwise it continues as normal. Not sure if I explained that correctly but hopefully you get the point

File: apps/phone/src/apps/twitter/components/tweet/TweetList.tsx (hunk 6 above)
Comment (GlitchOo) [reply apps/phone/src/apps/twitter/components/tweet/TweetList.tsx:19]: I see, I'm pretty sure this is easily solvable thinking about it. ill remove it for the meantime and set it back to the hardcoded limit of 25 and revisit this later. I have a pretty good idea how I could easily get around this problem.

File: apps/phone/src/apps/contacts/components/views/ContactInfo.tsx (hunk 8 above)
Comment (GlitchOo) [reply apps/phone/src/apps/contacts/components/views/ContactInfo.tsx:146]: You are correct, I meant to do this before pushing it but I guess I just moved on the next "to do"

File: .npmrc [hunk 9]
```diff
@@ -0,0 +1 @@
+public-hoist-pattern[]=*types*
```
[the remark this answers is not in the document]
Comment (GlitchOo) [reply .npmrc]: yeah I dont know to be honest, I think it happened when the npm for lodash was broken and I had to load it from github directly to get it to work for the build... it's removed though lol

File: apps/game/server/twitter/twitter.service.ts (hunk 1 above)
Comment (GlitchOo) [reply apps/game/server/twitter/twitter.service.ts]: changed

File: apps/phone/src/apps/contacts/hooks/useModal.tsx (hunk 3 above)
Comment (GlitchOo) [reply apps/phone/src/apps/contacts/hooks/useModal.tsx]: changed

File: apps/phone/src/apps/contacts/hooks/state.ts (hunk 4 above)
Comment (GlitchOo) [reply apps/phone/src/apps/contacts/hooks/state.ts]: 🤐

File: apps/phone/src/apps/twitter/components/AddTweetModal.tsx (hunk 5 above)
Comment (AvarianKnight) [reply apps/phone/src/apps/twitter/components/AddTweetModal.tsx:90]: shouldn't this just be `if (cleanedMessage.length > 0 && !isValidMessage(cleanedMessage)) return;` then?

The way this is formatted kinda just confuses me

File: apps/phone/src/index.tsx [hunk 10]
```diff
@@ -7,6 +7,7 @@ import dayjs from 'dayjs';
 import relativeTime from 'dayjs/plugin/relativeTime';
 import updateLocale from 'dayjs/plugin/updateLocale';
 import { Theme as MaterialUITheme } from '@mui/material';
+import { RewriteFrames } from '@sentry/integrations';
```
Comment (AvarianKnight) [inline apps/phone/src/index.tsx:10]: Unused import

File: apps/phone/src/os/apps/hooks/useApps.tsx [hunk 11]
```diff
@@ -8,6 +8,7 @@ import { useSettingsValue } from '../../../apps/settings/hooks/useSettings';
 import { IconSetObject } from '@typings/settings';
 import { useRecoilValue } from 'recoil';
 import { phoneState } from '@os/phone/hooks/state';
+import { extname } from 'path';
```
Comment (AvarianKnight) [inline apps/phone/src/os/apps/hooks/useApps.tsx:11]: Unused import

File: packages/database/src/darkchat/index.ts [hunk 12]
```diff
@@ -111,11 +111,7 @@ export class _DarkchatDB {
     return { id: result[0].id, label: result[0].label };
   }
 
-  async joinChannel(
-    channelIdentifier: string,
-    userIdentifier: string,
-    isOwner: boolean,
-  ): Promise<number> {
+  async joinChannel(channelIdentifier: string, userIdentifier: string, isOwner: boolean): Promise<number> {
```
Comment (AvarianKnight) [inline packages/database/src/darkchat/index.ts:114]: This shouldn't be collapsed, eslint show throw warnings/errors about this

File: typings/config.ts [hunk 13]
```diff
@@ -1,3 +1,5 @@
+import { Config } from "prettier";
```
Comment (AvarianKnight) [inline typings/config.ts:1]: Unused import

File: apps/game/client/functions.ts [hunk 14]
```diff
@@ -33,27 +33,17 @@ export const newPhoneProp = async () => {
     const [x, y, z] = GetEntityCoords(playerPed, true);
     global.phoneProp = CreateObject(GetHashKey(phoneModel), x, y, z + 0.2, true, true, true);
     const boneIndex = GetPedBoneIndex(playerPed, 28422);
-    AttachEntityToEntity(
```
Comment (AvarianKnight) [inline apps/game/client/functions.ts:36]: This shouldn't be collapsed, eslint should warn/error about this

**Review by AvarianKnight — CHANGES_REQUESTED** [review]
Some things I missed in my initial review

**Comment by AvarianKnight** [comment]
You also need to resolve the merge conficts

File: apps/phone/src/index.tsx (hunk 10 above)
Comment (GlitchOo) [reply apps/phone/src/index.tsx:10]: I never added this.. this was from the PR before mine. Ill remove it though as it shouldn't be there

File: apps/phone/src/os/apps/hooks/useApps.tsx (hunk 11 above)
Comment (GlitchOo) [reply apps/phone/src/os/apps/hooks/useApps.tsx:11]: Again same... never added this but ill remove

File: typings/config.ts (hunk 13 above)
Comment (GlitchOo) [reply typings/config.ts:1]: > You also need to resolve the merge conficts

i believe i can do that, however I am pretty sure they were caused by the PR before this

File: apps/game/client/functions.ts (hunk 14 above)
Comment (GlitchOo) [reply apps/game/client/functions.ts:36]: no warning or errors unless im just missing something

File: packages/database/src/darkchat/index.ts (hunk 12 above)
Comment (GlitchOo) [reply packages/database/src/darkchat/index.ts:114]: also no warnings or errors here as far as im aware. this is the first im ever hearing about this? was this a previous version because I upgraded eslint in the packages at some point im pretty sure

Comment (GlitchOo) [reply packages/database/src/darkchat/index.ts:114]: looking at much of the db class exports theyre all written in a similar fashion. is such a change necessary? if thats the case then shouldn't all of the classes be modified as well?

File: apps/game/client/functions.ts (hunk 14 above)
Comment (GlitchOo) [reply apps/game/client/functions.ts:36]: i made the change anyways, not sure what difference it actually makes

File: typings/config.ts (hunk 13 above)
Comment (GlitchOo) [reply typings/config.ts]: removed

File: apps/phone/src/os/apps/hooks/useApps.tsx (hunk 11 above)
Comment (GlitchOo) [reply apps/phone/src/os/apps/hooks/useApps.tsx]: removed

File: apps/phone/src/index.tsx (hunk 10 above)
Comment (GlitchOo) [reply apps/phone/src/index.tsx]: removed

**Comment by GlitchOo** [comment]
Okay I've resolved the conflicts... 188 conflicts in a pnpm lock file is on the list of one of the not so amazing things I've had to do lol

File: apps/phone/src/apps/twitter/components/AddTweetModal.tsx (hunk 5 above)
Comment (GlitchOo) [reply apps/phone/src/apps/twitter/components/AddTweetModal.tsx]: cleaned up

File: apps/phone/src/apps/twitter/components/AddTweetModal.tsx [hunk 15]
```diff
@@ -80,8 +80,14 @@ const AddTweetModal = () => {
   const submitTweet = async () => {
     await promiseTimeout(200);
     const cleanedMessage = clean(message.trim());
-    if (cleanedMessage.length === 0) return;
-    if (!isValidMessage(cleanedMessage)) return;
+    if (cleanedMessage.length === 0) {
```
Comment (itschip) [inline apps/phone/src/apps/twitter/components/AddTweetModal.tsx:83]: You can just check if `cleanedMessage` is false instead of the length :)

File: package.json [hunk 16]
```diff
@@ -40,5 +40,8 @@
     "hooks": {
       "pre-commit": "npx run format:staged"
     }
+  },
+  "dependencies": {
+    "lodash": "https://github.com/lodash/lodash.git#npm"
```
Comment (itschip) [inline package.json:45]: This shouldn't be here. Uninstall it, and install it in the package where it belong :)

**Review by itschip — CHANGES_REQUESTED** [review]

File: apps/phone/src/apps/twitter/components/AddTweetModal.tsx (hunk 15 above)
Comment (GlitchOo) [reply apps/phone/src/apps/twitter/components/AddTweetModal.tsx:83]: yeah I suppose i just left what was there from whoever wrote the code. ill change it now and push

File: package.json (hunk 16 above)
Comment (GlitchOo) [reply package.json:45]: yeah think this happened when i was having issues with getting the correct lodash ill clean that up now

Comment (GlitchOo) [reply package.json]: fixed

File: apps/phone/src/apps/twitter/components/AddTweetModal.tsx (hunk 15 above)
Comment (GlitchOo) [reply apps/phone/src/apps/twitter/components/AddTweetModal.tsx]: cleaned up, should be better

File: package.json [hunk 17]
```diff
@@ -16,7 +16,8 @@
     "rollup": "^2.77.2",
     "rollup-plugin-dts": "^4.2.2",
     "rollup-plugin-esbuild": "^4.10.2",
-    "turbo": "^1.8.3"
+    "turbo": "^1.8.3",
+    "lodash": "https://github.com/lodash/lodash.git#npm"
```
Comment (itschip) [inline package.json:20]: What are you using lodash for? I don't see any usage with it. 

and if it's only used in `apps/phone`, it should be installed in that package.json

**Review by itschip — CHANGES_REQUESTED** [review]

File: package.json (hunk 17 above)
Comment (GlitchOo) [reply package.json:20]: It's a dependency for eslint and a few other packages in the dev dependency. The dependency downloaded by npn is broken and doesn't have the modules required so the build will just fail

**Review by itschip — APPROVED** [review]

**Review by AvarianKnight — APPROVED** [review]
LGTM

**Comment by TasoOneAsia** [comment]
Taking a look at this a little bit after the fact and want to reiterate good work :) Not often do I see such a high-quality PR

**Comment by TonybynMp4** [comment]
damn the messiah complimented you that's.. something 😲

**Comment by GlitchOo** [comment]
> Taking a look at this a little bit after the fact and want to reiterate good work :) Not often do I see such a high-quality PR

I appreciate that, thank you
